javascript_nodejs/17.proactive-messages/dialogs/mainDialog/index.js

```diff
@@ -90,40 +100,60 @@ class MainDialog {
+        let jobInfo = jobs[jobIdNumber];
 
         // If no job was found, notify the user of this error state.
-        if (isEmpty(jobInfo)){
+        if (isEmpty(jobInfo)) {
             await turnContext.sendActivity(`Sorry no job with ID ${ jobIdNumber }.`);
         } else {
             // Found a job with the ID passed in.
-            const reference = jobInfo[jobIdNumber].reference;
-            const completed = jobInfo[jobIdNumber].completed;
+            const reference = jobInfo.reference;
+            const completed = jobInfo.completed;
 
             // If the job is not yet completed and conversation reference exists,
             // use the adapter to continue the conversation with the job's original creator.
             if (reference && !completed) {
                 // Since we are going to proactively send a message to the user who started the job,
                 // we need to create the turnContext based on the stored reference value.
-                await this.adapter.continueConversation(reference, async (turnContext) => {
+                await this.adapter.continueConversation(reference, async (proactiveTurnContext) => {
                     // Complete the job.
-                    jobInfo[jobIdNumber].completed = true;
+                    jobInfo.completed = true;
                     // Save the updated job.
-                    await storage.write(jobInfo);
+                    await this.jobsList.set(turnContext, jobs);
                     // Notify the user that the job is complete.
-                    await turnContext.sendActivity('Your queued job just completed.');
+                    await proactiveTurnContext.sendActivity('Your queued job just completed.');
                 });
+
+                // Send a message to the person who completed the job.
+                await turnContext.sendActivity('Job completed. Notification sent.');
+
             }
             // The job has already been completed.
-            else if (reference && completed) {
-                await this.adapter.continueConversation(reference, async (turnContext) => {
-                    await turnContext.sendActivity('This job is already completed, please start a new job.');
-                });
+            else if (completed) {
+                await turnContext.sendActivity('This job is already completed, please start a new job.');
             };
         };
     };
+
+    // Show a list of the pending jobs
+    async showJobs(turnContext) {
+        const jobs = await this.jobsList.get(turnContext, {});
+        if (Object.keys(jobs).length) {
+            await turnContext.sendActivity(
+                "| Job number &nbsp; | Conversation ID &nbsp; | Completed |<br>" +
+                "| :--- | :---: | :---: |<br>" +
+                Object.keys(jobs).map((key) => {
+                    return `${ key } &nbsp; | ${ jobs[key].reference.activityId } &nbsp; | ${ jobs[key].completed }`
```

Comment:
Please change to:

```javascript
return `${ key } &nbsp; | ${ jobs[key].reference.conversation.id.split('|')[0] } &nbsp; | ${ jobs[key].completed }`
```

The important information is the conversation ID, not the activity ID. The ugly `.split` at the end is to not include a pipe character in the output, which breaks the markdown formatting.